vscode/src/notebook.ts

```diff
@@ -65,6 +72,98 @@ export function registerQSharpNotebookHandlers() {
   return subscriptions;
 }
 
+const openQSharpNotebooks = new Set<string>();
+
+/**
+ * This one is for syncing with the language service
+ */
+export function registerQSharpNotebookCellUpdateHandlers(
+  languageService: ILanguageService,
+) {
+  vscode.workspace.notebookDocuments.forEach((notebook) => {
+    updateIfQsharpNotebook(notebook);
+  });
+
+  const subscriptions = [];
+  subscriptions.push(
+    vscode.workspace.onDidOpenNotebookDocument((notebook) => {
+      updateIfQsharpNotebook(notebook);
+    }),
+  );
+
+  subscriptions.push(
+    vscode.workspace.onDidChangeNotebookDocument((event) => {
+      updateIfQsharpNotebook(event.notebook);
+    }),
+  );
+
+  subscriptions.push(
+    vscode.workspace.onDidCloseNotebookDocument((notebook) => {
+      closeIfKnownQsharpNotebook(notebook);
+    }),
+  );
+
+  function updateIfQsharpNotebook(notebook: vscode.NotebookDocument) {
+    if (notebook.notebookType === jupyterNotebookType) {
+      const qsharpCells = getQSharpCells(notebook);
+      const notebookUri = notebook.uri.toString();
+      if (qsharpCells.length > 0) {
+        openQSharpNotebooks.add(notebookUri);
+        languageService.updateNotebookDocument(
+          notebookUri,
+          notebook.version,
+          qsharpCells.map((cell) => {
+            return {
+              uri: cell.document.uri.toString(),
+              version: cell.document.version,
+              code: getQSharpText(cell.document),
+            };
+          }),
+        );
+      } else {
+        // All Q# cells could have been deleted, check if we know this doc from previous calls
+        closeIfKnownQsharpNotebook(notebook);
+      }
+    }
+  }
+
+  function closeIfKnownQsharpNotebook(notebook: vscode.NotebookDocument) {
+    const notebookUri = notebook.uri.toString();
+    if (openQSharpNotebooks.has(notebookUri)) {
+      languageService.closeNotebookDocument(
+        notebookUri,
+        getQSharpCells(notebook).map((cell) => cell.document.uri.toString()),
+      );
+      openQSharpNotebooks.delete(notebook.uri.toString());
+    }
+  }
+
+  function getQSharpCells(notebook: vscode.NotebookDocument) {
+    return notebook
+      .getCells()
+      .filter((cell) =>
+        vscode.languages.match(qsharpDocumentFilter, cell.document),
+      );
+  }
+
+  function getQSharpText(document: vscode.TextDocument) {
+    // Erase the %%qsharp magic line if it's there
+    // Replace it with whitespace so that document offsets remain the same.
+    // This will save us from having to map offsets later when
+    // communicating with the language service.
+    if (document.lineAt(0).text.startsWith(qsharpCellMagic)) {
```

Comment:
Is it common practice to require language magic to be the very first line in a a cell?

Reply:
Are you asking about allowing whitespace/empty lines before the cell magic? Technically, IPython requires the cell magic to be the first thing in the cell, it does allow empty lines / whitespace before it. Though I've never seen any user code that doesn't just put the cell magic on the first line.

We don't skip whitespace here, so we're being more restrictive than IPython, but there's no real reason for it... I can update our code to allow whitespace.